pennylane/math/utils.py

```diff
@@ -524,3 +530,116 @@ def in_backprop(tensor, interface=None):
         return False
 
     raise ValueError(f"Cannot determine if {tensor} is in backpropagation.")
+
+
+def pwc_from_array(dt, index):
+    """Create a function that is piecewise-constant in time, based on the params for a TDHamiltonian.
+
+    Args:
+        dt(Union[float, tuple(float, float)]: the total duration as a float, or the start and end time as floats.
+        index(int): the index at which the relevant parameter array is located in the overall ``params`` variable
+
+    Returns:
+        func: a function that can be passed the full ``params`` variable and ``t``, and will return the
+            corresponding constant
+
+    **Example**
+
+    >>> t1, t2 = 1, 3
+    >>> index = 0
+    >>> f1 = pwc_from_array((t1, t2), index)
+
+    The resulting function ``f1`` has the call signature ``f1(params, t)``. If passed parameters and a time,
+    it will assign the array at ``params[index]`` as the constants in the piecewise function, and select
+    the constant corresponding to the specified time, based on the previously specified interval ``t1`` to ``t2``.
+
+    >>> params = [np.linspace(10, 20, 10)]
+    >>> f1(params, 2)
+    tensor(15.55555556, requires_grad=True)
+
+    >>> f1(params, 2.1)  # same bin
+    tensor(15.55555556, requires_grad=True)
+
+    >>> f1(params, 2.5)  # next bin
+    tensor(17.77777778, requires_grad=True)
+    """
+
+    if isinstance(dt, tuple):
+        t1, t2 = dt
+    else:
+        t1 = 0
+        t2 = dt
+
+    def func(params, t):
+        params = params[index]
+        num_bins = len(params)
+        if has_jnp:
+            idx = jnp.array(num_bins / (t2 - t1) * (t - t1), dtype=int)
+            # ToDo: also base this on get_interface, or is it fine for idx to always be a jax array if jax is available?
+        else:
+            idx = np.array(num_bins / (t2 - t1) * (t - t1), dtype=int)  # not jittable
+        return params[idx]
+
+    return func
+
+
+def pwc_from_function(dt, num_bins):

```

Comment:
Interesting! I havent even thought of that possibility. Did you have a specific use case in mind for this?